Updating README.md (#46)
VirtusLab/akka-serialization-helper: README.md

```scala
enablePlugins(DumpEventSchemaPlugin)
```

You also have to change one setting responsible for resolving companion compiler plugin:

```scala
dumpEventSchema / dumpEventSchemaCompilerPlugin := "com.github.VirtusLab.akka-serialization-helper" % "dump-event-schema-compiler-plugin" % commit 
```

### 3. Serializer

Simple [borer-based](https://github.com/sirthias/borer) Akka serializer. It uses codecs, provided by Borer, that are
generated during compile time (so serializer won't crash during runtime like reflection-based serializers may do).

```scala
libraryDependencies += "com.github.VirtusLab.akka-serialization-helper" %% "borer-akka-serializer" % commit
```

It may also be worth including additional codecs for common types that are missing in Borer standard library:

```scala
libraryDependencies += "com.github.VirtusLab.akka-serialization-helper" %% "borer-extra-codecs" % commit
```

## Comparison with other Akka Serializers

| Serializer | Jackson | Circe | Protobuf v3 | Avro | Borer | Kryo
|:---|:---|:---|:---|:---|:---|:---|
| Data formats | JSON or [CBOR](https://cbor.io) | JSON | JSON or binary | JSON or binary | JSON or CBOR | binary

> | Scala support | with [`jackson-module-scala`](https://github.com/FasterXML/jackson-module-scala) <ul><li>lacks support of basic scala types like `Unit`</li><li>without explicit annotation doesn't work with generics extending `AnyVal`</ul> | perfect | with [`ScalaPB`](https://scalapb.github.io) generates Scala ADTs based on protobuf definitions | with [`Avro4s`](https://github.com/sksamuel/avro4s) library generaters Avro schema based on Scala ADTs | perfect | perfect

Reviewer: I'd even put the fact the Jackson lacks support for Scala object as primary objection here (rather than in `Boilerplate` row)

| Akka support | [library](https://doc.akka.io/docs/akka/current/serialization-jackson.html) | requires custom serializer | [akka.remote](https://doc.akka.io/docs/akka/current/serialization.html) | requires custom serializer | requires custom serializer | [library](https://github.com/altoo-ag/akka-kryo-serialization)

> | Runtime safety | none, uses reflection | encoders and decoders are checked during compile time | standard for scala code | standard for scala code | encoders and decoders are checked during compile time | none, uses reflections

Reviewer: Nit:
```suggestion
| Runtime safety | none, uses reflection | encoders and decoders are checked during compile time | standard for scala code | standard for scala code | encoders and decoders are checked during compile time | none, uses reflection
```

Reviewer: Also... for Protobuf, where is that `akka.remote` name coming from? 🤔 akka-remote library itself is typically written with a hyphen

Author: Serializer comes from `akka-remote`

Reviewer: Ok so pls change to `akka-remote` with hyphen then ➖

> | Schema evolution | <ul><li>removing field</li><li>adding optional field</li></ul> with [`JacksonMigration`](https://doc.akka.io/docs/akka/current/serialization-jackson.html#schema-evolution) <ul><li>adding mandatory field</li><li>renaming field</li><li>renaming class</li><li>support of forward versioning for rolling updates</li></ul>| <ul><li>adding optional field</li><li>removing optional field</li><li>adding required field with default value</li><li>removing required field</li><li>renaming field</li><li>reordering fields</li><li>transforming data before deserialization</li></ul> | <ul><li>adding optional field</li><li>removing optional field</li><li>adding required field with default value</li><li>removing required field</li><li>renaming field</li><li>reordering fields</li><li>changing between compatible types</li></ul>  | <ul><li>reordering fields</li><li>renaming fields</li><li>adding optional field</li><li>adding required field with default value</li><li>removing field with default value</li></ul> | <ul><li>renaming fields</li><li>transforming data before deserialization</li></ul> | <ul><li>adding field</li><li>removing field</li><li>renaming field</li></ul>

Reviewer: Ok... so now we're listing **everything** that's possible, or only the stuff that's available OOTB/with low effort?

Reviewer: Maybe even having two separate rows for each category won't be an overkill here (?)

Author: > Maybe even having two separate rows for each category won't be an overkill here
I think it would be overkill, because only Borer and Circe have that functionality

Author: > Ok... so now we're listing everything that's possible, or only the stuff that's available OOTB/with low effort?

Everything

Reviewer: Wow okay so with Kryo it's even impossible to **rename a class** right? 😯

Author: From my understanding of Kryo, it uses registration ID instead of class name, so class names don't matter

> | Boilerplate | a lot: <ul><li>ADTs requires amount of annotation equal to or exceeding the actual type definitions</li><li>poor support for case object - requires custom serializer and deserializer and sealed trait per case object, otherwise, it creates new singletons and breaks pattern matching</ul> | a bit: every top level sealed trait must be mentioned for Circe to generate a codec | in case of custom types a second layer of models is needed | sometimes requires annotations | considerable: every top level sealed trait must have manually defined codec | none

Reviewer: Hmmm Circe _a bit_ but Borer _considerable_... but the actual description looks almost identical 🤔

Reviewer: I suppose the effort is considerable for Borer b/c you need to define codecs for transitively included types, while Circe relies on Magnolia for that, right?

Author: Yes, Circe relies on Magnolia.
Also Borer has separate codec[A] definition and register[A] call.

```scala
resolvers += "jitpack".at("https://jitpack.io")

> val commit = "master-SNAPSHOT" //name of a branch or commit

Reviewer: Mention that for commit hash, no `-SNAPSHOT` is expected (at least that wouldn't make sense in principle 🤔 )

The project consists of three modules that are independent of each other, comprising together a complete solution.

### 1. Checker Plugin

A Scala compiler plugin that detects messages, events etc. and checks, whether they extend the base trait and reports

> and error when it doesn't. This ensures that the specified serializer is used by Akka and protects against accidental use

Reviewer: Link to akka docs on why Java serialization is discouraged

> The project consists of three modules that are independent of each other, comprising together a complete solution.

Reviewer: ```suggestion
The project consists of three modules that are independent of each other, comprising a complete solution together .
```

Dunno why, just sounds better to me ;)

### 1. Checker Plugin

> A Scala compiler plugin that detects messages, events etc. and checks, whether they extend the base trait and reports

Reviewer: ```suggestion
A Scala compiler plugin that detects messages, events and persistent state, and checks whether they extend the base trait and reports
```

plus a link to each of 3 categories (esp. given that it's the first they're mentioned)

Author: Not sure I there is a link to add. These terms are defined in documentation in an offhand manner. Links I provide will seem unrelated at the first glance.

use of [Java serialization](https://doc.akka.io/docs/akka/current/serialization.html#java-serialization).

To use, just annotate a base trait with `@org.virtuslab.ash.SerializabilityTrait`:

```scala
@SerializabilityTrait
trait MySerializable

```

Installation:

```scala
libraryDependencies += "com.github.VirtusLab.akka-serialization-helper" %% "serializability-checker-library" % commit
libraryDependencies += compilerPlugin("com.github.VirtusLab.akka-serialization-helper" %% "serializability-checker-compiler-plugin" % commit)
```

### 2. Dump Event Schema

An sbt plugin that allows for dumping schema
of [akka-persistence](https://doc.akka.io/docs/akka/current/typed/persistence.html#example-and-core-api) events to a
file. Can be used for detecting accidental changes of events.

Installation:

Add to `plugins.sbt`:

```scala
libraryDependencies += "com.github.VirtusLab.akka-serialization-helper" % "sbt-dump-event-schema" % commit
```

To enable the plugin for a specific project, use:

```scala
enablePlugins(DumpEventSchemaPlugin)
```

You also have to change one setting responsible for resolving companion compiler plugin:

```scala

> dumpEventSchema / dumpEventSchemaCompilerPlugin := "com.github.VirtusLab.akka-serialization-helper" % "dump-event-schema-compiler-plugin" % commit 

Reviewer: Hmm would it be possible that this compiler plugin is automatically resolved to jitpack and not to a local snapshot? Or will it only be reasonably doable once we move to Maven Central? (curious how it's handled by sbt plugins in general)

Author: Not really, that would require an sbt plugin to know how was it resolved

Reviewer: Ok... I suppose that's the answer to first part of the question (jitpack)? or to both questions (jitpack + maven centrl)?

Author: With maven I solve this with default value:
```scala
dumpEventSchema / dumpEventSchemaCompilerPlugin := "org.virtuslab" %% "dump-event-schema-compiler-plugin" % "0.1.0-SNAPSHOT"
```

Reviewer: Ok, that's what I suspected... I guess with Maven we'd need to somehow make sure that the default value for `dumpEventSchemaCompilerPlugin` will point to same version of the compiler plugin as the version of sbt plugin itself... as we rather don't want to maintain two independent versioning schemes for sbt plugin and compiler plugin :/

```

### 3. Serializer

> Simple [Borer](https://github.com/sirthias/borer) based Akka serializer. It uses codecs, provided by Borer, that are

Reviewer: nit:
```suggestion
Simple [Borer](https://github.com/sirthias/borer)-based Akka serializer. It uses codecs, provided by Borer, that are
```

generated during compile time (so serializer won't crash during runtime like reflection-based serializers may do).

```scala
libraryDependencies += "com.github.VirtusLab.akka-serialization-helper" %% "borer-akka-serializer" % commit
```

It may also be worth including additional codecs for common types that are missing in Borer standard library:

```scala
libraryDependencies += "com.github.VirtusLab.akka-serialization-helper" %% "borer-extra-codecs" % commit
```

## Comparison with other Akka Serializers

> | Serializer | Jackson | Circe | Protobuf v3 | Avro | Borer |

Reviewer: Pls add a link for each serializer (not just link to the main page of the given library, but rather to the page specifically dedicated to Akka serialization)

|:---|:---|:---|:---|:---|:---|

> | Data formats | JSON or CBOR | JSON | JSON or binary | JSON or binary | JSON or CBOR |

Reviewer: Pls link to CBOR since it's not as widely known as JSON (I've only learnt of its existence in Hydra)

| Scala support | with [`jackson-module-scala`](https://github.com/FasterXML/jackson-module-scala) <ul><li>lacks support of basic scala types like `Unit`</li><li>without explicit annotation doesn't work with generics extending `AnyVal`</ul> | perfect | with [`ScalaPB`](https://scalapb.github.io) generates Scala ADTs based on protobuf definitions | with [`Avro4s`](https://github.com/sksamuel/avro4s) library generaters Avro schema based on Scala ADTs | perfect

> | Akka support | out of the box | requires custom serializer | requires custom serializer | requires custom serializer | requires custom serializer |

Reviewer: Wow really other than Jackson, none of them ships with a ready Akka serializer? 😯 (Kryo does for sure tho)

Reviewer: As in, maybe not _ships with_ but _has a ready library somewhere out there on github that provides_

> | Runtime safety | none <ul><li>uses reflection</li><li>errors appear only in runtime</li></ul> | encoders and decoders are checked during compile time | generates scala classes | standard for scala code | encoders and decoders are checked during compile time

Reviewer: Hmm for protobuf, `generates scala classes` doesn't really fit into `Runtime safety` category... probably `standard for scala code` (as for Avro) makes more sense

> | Schema evolution | <ul><li>removing field</li><li>adding optional field</li></ul> with [`JacksonMigration`](https://doc.akka.io/docs/akka/current/serialization-jackson.html#schema-evolution) <ul><li>adding mandatory field</li><li>renaming field</li><li>renaming class</li><li>support of forward versioning for rolling updates</li></ul>| <ul><li>adding optional field</li><li>removing optional field</li><li>adding required field with default value</li><li>removing required field</li><li>renaming field</li><li>reordering fields</li></ul> | <ul><li>switching between optional and repeated field</li><li>adding new fields</li><li>renaming fields</li></ul> | <ul><li>reordering fields</li><li>renaming fields</li><li>adding optional field</li><li>adding required field with default value</li><li>removing field with default value</li></ul> | only renaming fields

Reviewer: Hmm for Borer: _only renaming fields_ refers to what's provided out of the box... but all other operations are possible as well, right?

Reviewer: I suppose that's the case for all other libs: that the row lists what is provided OOTB rather than what's possible in principle (?)... feel free to contact the original author(s) of this table if needed

> | Boilerplate | a lot: <ul><li>ADTs requires amount of annotation equal to or exceeding the actual type definitions</li><li>poor support for case object - requires custom serializer and deserializer and sealed trait per case object, otherwise, it creates new singletons and breaks pattern matching</ul> | a bit: every top level sealed trait must be mentioned for Circe to generate a codec | in case of custom types a second layer of models is needed | sometimes requires annotations | considerable: every top level sealed trait must have manually defined codec

Reviewer: I'd rather put `considerable:`  before Protobuf, since relative to Borer it's a ton of code (you need to provide .proto files for each type, right?)

Author: Not really. After creating Protobuf file, scala classes are generated from it. So there is no model duplication.

Reviewer: Ok, right, makes sense, the amount of effort is similar

Installation:

```scala
libraryDependencies += "com.github.VirtusLab.akka-safer-serializer" %% "akka-serializability-checker-library" % commit
libraryDependencies += compilerPlugin("com.github.VirtusLab.akka-safer-serializer" %% "akka-serializability-checker-compiler-plugin" % commit)
```

### 2. Dump Event Schema

> An sbt plugin that allows for dumping schema of events to a file. Can be used for detecting accidental changes of

Reviewer: Mention that you mean akka-persistent events, with some relevant link to Akka docs

### 2. Dump Event Schema

> A sbt plugin that allows for dumping schema of events to a file. Can be used for detecting accidental changes of

Reviewer: An es-bee-tee, was good ;)

events.

Installation:

Add to `plugins.sbt`:
```scala

> libraryDependencies += repo % "sbt-dump-event-schema" % commit

Reviewer: Actually, I'd repeat the jitpack'd organization name explictly everywhere instead of using `repo` (also, this should've been `org`, not `repo`) for clarity

Author: Done it, not sure if it improved clarity. Because of increased length some code snippets have acquired navigation bar.

Author: @PawelLipski

Reviewer: Okay, I still like the version with verbatim org name more, at least users don't need to look around the README for the missing pieces as much

Add to `plugins.sbt`:
```scala
libraryDependencies += "com.github.VirtusLab.akka-safer-serializer" % "sbt-dump-event-schema" % commit
```
To enable the plugin for a specific project, use:
```scala
enablePlugins(DumpEventSchemaPlugin)
```
You also have to change one setting responsible for resolving companion compiler plugin:
```scala
dumpEventSchema / dumpEventSchemaCompilerPlugin := "com.github.VirtusLab.akka-safer-serializer" % "dump-event-schema-compiler-plugin" % commit 
```

### 3. Serializer

> Simple Borer based Akka serializer. It uses codecs, provided by Borer, that are generated during compile time (so

Reviewer: Link to borer

```

## Comparison with other Akka Serializers

| Serializer | Jackson | Circe | Protobuf v3 | Avro | Borer |
|:---|:---|:---|:---|:---|:---|
| Data formats | JSON or CBOR | JSON | JSON or binary | JSON or binary | JSON or CBOR |
| Scala support | with [`jackson-module-scala`](https://github.com/FasterXML/jackson-module-scala) <ul><li>lacks support of basic scala types like `Unit`</li><li>without explicit annotation doesn't work with generics extending `AnyVal`</ul> | perfect | with [`ScalaPB`](https://scalapb.github.io) generates Scala ADTs based on protobuf definitions | with [`Avro4s`](https://github.com/sksamuel/avro4s) library generaters Avro schema based on Scala ADTs | perfect
| Akka support | out of the box | requires custom serializer | requires custom serializer | requires custom serializer | requires custom serializer |
| Runtime safety | none <ul><li>uses reflection</li><li>errors appear only in runtime</li></ul> | encoders and decoders are checked during compile time | generates scala classes | standard for scala code | encoders and decoders are checked during compile time

> | Schema evolution | <ul><li>removing field</li><li>adding optional field</li></ul> with [`JacksonMigration`](https://doc.akka.io/docs/akka/current/serialization-jackson.html#schema-evolution) <ul><li>adding mandatory field</li><li>renaming field</li><li>renaming class</li><li>support of forward versioning for rolling updates</li></ul>| <ul><li>adding optional field</li><li>removing optional field</li><li>adding required field with default value</li><li>removing required field</li><li>renaming field</li><li>reordering fields</li></ul> | <ul><li>switching between optional and repeated field</li><li>adding new fields</li><li>renaming fields</li></ul> | <ul><li>reordering fields</li><li>renaming fields</li><li>adding optional field</li><li>adding required field with default value</li><li>removing field with default value</li></ul> | only renaming fields

Reviewer: Is it possible to declare a table row in markdown in multiple lines? In fact, I think even raw HTML might be more readable/less error-prone here

Author: Not really.
Raw HTML doesn't have styling that MD table has, so it looks awful.

Reviewer: Ok ☹️ let's leave as is then

| Serializer | Jackson | Circe | Protobuf v3 | Avro | Borer | Kryo
|:---|:---|:---|:---|:---|:---|:---|
| Data formats | JSON or [CBOR](https://cbor.io) | JSON | JSON or binary | JSON or binary | JSON or CBOR | binary

> | Scala support | very poor, even with [`jackson-module-scala`](https://github.com/FasterXML/jackson-module-scala) <ul><li>lacks support of basic scala types like `Unit`</li><li>without explicit annotation doesn't work with generics extending `AnyVal`</li><li>poor support for case object, without configuration creates new singletons, breaking pattern matching</li></ul> | perfect | with [`ScalaPB`](https://scalapb.github.io) generates Scala ADTs based on protobuf definitions | with [`Avro4s`](https://github.com/sksamuel/avro4s) library generaters Avro schema based on Scala ADTs | perfect | perfect

Reviewer: Nit:
```suggestion
| Scala support | very poor, even with [`jackson-module-scala`](https://github.com/FasterXML/jackson-module-scala) <ul><li>lacks support of basic scala types like `Unit`</li><li>without explicit annotation doesn't work with generics extending `AnyVal`</li><li>poor support for Scala objects, without configuration creates new singletons, breaking pattern matching</li></ul> | perfect | with [`ScalaPB`](https://scalapb.github.io) generates Scala ADTs based on protobuf definitions | with [`Avro4s`](https://github.com/sksamuel/avro4s) library generaters Avro schema based on Scala ADTs | perfect | perfect
```

Coz it's a problem with non-case objects AFAIK as well

> | Boilerplate | a lot: <ul><li>ADTs requires amount of annotation equal to or exceeding the actual type definitions</li><li>requires custom serializer and deserializer and sealed trait per case object</ul> | every top level sealed trait must be mentioned for Circe to generate a codec | in case of custom types a second layer of models is needed | sometimes requires annotations | every top level sealed trait must have manually defined codec | every class must be passed to register method with unique id

Reviewer: Nit:

```suggestion
| Boilerplate | a lot: <ul><li>ADTs requires amount of annotation equal to or exceeding the actual type definitions</li><li>requires custom serializer and deserializer and sealed trait per case object</ul> | every top level sealed trait must be mentioned for Circe to generate a codec | in case of custom types a second layer of models is needed | sometimes requires annotations | every top level sealed trait must have manually defined codec | every class must be passed to `register` method with unique id
```

I suppose that's what you meant (?)
